src/FwdState.cc

```diff
@@ -873,6 +873,15 @@ FwdState::noteConnection(HappyConnOpener::Answer &answer)
 
     if (error) {
         fail(error);
+        if (request->clientConnectionManager->port->flags.tunnelSslBumping
+            && request->clientConnectionManager->port->secure.terminateOnSecureConnectFail
+            && error->type == ERR_CONNECT_FAIL) {
+            
+            flags.dont_retry = true;
+            entry->abort();
+            complete();
+            return;
+        }
```

Comment:
After testing this PR a little, I would like to add the following to my blocking [review](https://github.com/squid-cache/squid/pull/1668#discussion_r1486304646): The current PR code does result in "HTTPS to HTTP" downgrade in Chrome, as intended, but the desired Chrome behavior appears to be triggered by a _timeout_ in Chrome (i.e. by Squid inaction): In my quick-and-dirty tests, The original CONNECT tunnels opened by Chrome remain open when Chrome (after a ~6 second pause) sends a plain text request to Squid. If that observation is not a side effect of my poor testing, then I hope Squid behavior can be changed when PR code is refactored to address earlier review concerns: If it all possible, we want Squid to close doomed CONNECT tunnels; we do not want these tunnels to wait for browser timeouts.

Reply:
I am not sure what is actually causing the difference though, the browser does not seem to (generally) wait the timeout on my environment.

I am aware of that the timeout is one of the triggers of HTTPS to HTTP downgrade. If the upstream HTTPS port is unreachable (TCP SYN is silently dropped), Squid wait TCP ACK for a certain amount of time then that perhaps results in the same behaviour of browser like you observe. Does it match to your test case?

Reply:
Note: if waiting timeout generally occurs depending on the platform, sending TCP RST (setsockopt) at the disconnection may be an idea to try. I prototyped before posting this PR but did not include it there because I did not find any benefit on my environment.

Reply:
> I am not sure what is actually causing the difference though, the browser does not seem to (generally) wait the timeout on my environment.

What signal/information do you think your browser is acting on if it is not the timeout? AFAICT, Squid sends nothing (but a TCP ACK packet) to the browser after responding to CONNECT... In your tests, does Squid close the browser-to-Squid connection before browser downgrades?

> If the upstream HTTPS port is unreachable (TCP SYN is silently dropped), Squid wait TCP ACK for a certain amount of time then that perhaps results in the same behaviour of browser like you observe. Does it match to your test case?

Please note that what happens between Squid and the origin server is irrelevant to my concern. Said that, IIRC, in my test, Squid "immediately" receives a RST packet because nobody is listening on the HTTPS port of the test origin server.

Reply:
> What signal/information do you think your browser is acting on if it is not the timeout? AFAICT, Squid sends nothing (but a TCP ACK packet) to the browser after responding to CONNECT... In your tests, does Squid close the browser-to-Squid connection before browser downgrades? 


Edge: CONNECT tcp session is immediately (of course with slight delay to access the origin server) disconnected by FIN, and GET follows it)
![https-upgrade-edge](https://github.com/squid-cache/squid/assets/145020979/ced6a2ae-9d4b-4edb-b73b-d0e9c4fd93ce)


Chrome: needed to push the security warning button before actually access to the downgraded HTTP, so there is a little interval between CONNECT and GET, but you will see CONNECT tcp session is immediately disconnected by FIN.
![https-upgrade-chrome](https://github.com/squid-cache/squid/assets/145020979/23dd9d14-3a76-4470-a0d9-192ec267b029)

Reply:
My test environment is also dirty.
I'm using the version of v6.6 with KTLS patch (and some more personal patch like increasing buffer size)..
If the difference persists, we have to narrow down the cause.

Reply:
> If the difference persists, we have to narrow down the cause.

Agreed. I am not particularly worried about kept-open browser-Squid connections at this time as long as we agree that they should not be kept open. We will retest when other (primary) concerns are addressed.

Reply:
This seems to be the one.
> a side effect of a different ssl_bump configuration (I tested with ssl_bump stare all) or some other valid/reasonable but different configuration options

I had been using ssl_bump bump all.
Changing it to ssl_bump stare all reprocuces the timeout.

Reply:
Changed the code to proactively terminate (reset) tcp as of https://github.com/squid-cache/squid/pull/1668/commits/1943d24c41fecff8cf46879c28dac0e75881b8b0.